### PR TITLE
Broadcast Service Event Queue

### PR DESCRIPTION
**Pull request by xTEddie**
### Description
In rare cases, broadcast events get dropped for unknown reason.

## Solution Proposed
We're adding a retry mechanism in case the broadcast events get dropped.

### Acceptance criteria
Event queue is working as expected

## Test cases and evidence
_Include what tests cases were considered, any evidence of testing for future references, to identify any corner cases, etc_

### Sanity Tests
-   [ ] You have tested all changes in Popout mode
-   [ ] You have tested all changes in cross browsers i.e Edge, Chrome, Firefox, Safari and mobile devices(iOS and Android)
-   [ ] Your changes are included in the [CHANGELOG](../CHANGE_LOG.md)

## A11y
- [ ] You have run the [Automated Accessibility Check](https://accessibilityinsights.io/docs/en/windows/getstarted/automatedchecks) and have no reported issues

__*Please provide justification if any of the validations has been skipped.*__

### CODE REVIEW

File: chat-components/src/services/BroadcastService.ts [hunk 1]
```diff
@@ -45,7 +104,16 @@ export const BroadcastService = {
...
-        pubChannel.postMessage(JSON.parse(JSON.stringify(message)));
+        try {
+            const messageCopy = JSON.parse(JSON.stringify(message));
+            const eventId = new Date().getTime();
+            const event = {...messageCopy, eventId};
+
+            pubChannel.postMessage(event);
+            eventQueue.pushEvent(event);
+        } catch (error) {
+            console.error("Error in BroadcastService.postMessage:", error);
```
Comment (elopezanaya) [inline chat-components/src/services/BroadcastService.ts:115]: we need log this in telemetry

File: chat-components/src/services/BroadcastService.ts [hunk 2]
```diff
@@ -14,8 +14,66 @@ const broadcastServiceSubList: Record<string, any> = {};
 let pubChannel: any;
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
 let subChannel: any;
+let eventQueue: EventQueue;
+
+class EventQueue {
```
Comment (elopezanaya) [inline chat-components/src/services/BroadcastService.ts:19]: move it outisde this class

File: chat-components/src/services/BroadcastService.ts [hunk 3]
```diff
@@ -14,8 +14,66 @@ const broadcastServiceSubList: Record<string, any> = {};
...
+    stopQueueing() {
+        this.queueing = false;
+        this.queuingId = undefined
+    }
+
+    pushEvent(event: any) {
+        this.channelEventQueue.push(event);
+    }
+
+    popEvent(event: any) {
```
Comment (elopezanaya) [inline chat-components/src/services/BroadcastService.ts:60]: lets use a Map<Integer, Object> so we access to the id in O(1) instead of filtering O(n)

File: chat-components/src/services/BroadcastService.ts [hunk 4]
```diff
@@ -14,8 +14,66 @@ const broadcastServiceSubList: Record<string, any> = {};
...
+    }
+
+    queueEvents(timeout = 3000) {
+        this.processEvents();
+        if (this.queueing) {
+            this.queuingId = setTimeout(() => {
+                this.queueEvents();
+            }, timeout);
+        }
+    }
```
Comment (elopezanaya) [inline chat-components/src/services/BroadcastService.ts:45]: I suggest that instead of waiting 3 secs, to react immediately with 3 retries, 500 ms between retries, 


this aggressive approach aims to immediate notification in this fallback mechanism

File: chat-components/src/services/BroadcastService.ts [hunk 5]
```diff
@@ -45,7 +104,16 @@ export const BroadcastService = {
          * Omit copying methods to prevent 'DataCloneError' in older browsers when passing an object with functions
          * This exception occurs when an object can't be clone with the 'structured clone algorithm' (used by postMessage)
          */
-        pubChannel.postMessage(JSON.parse(JSON.stringify(message)));
+        try {
+            const messageCopy = JSON.parse(JSON.stringify(message));
+            const eventId = new Date().getTime();
```
Comment (elopezanaya) [inline chat-components/src/services/BroadcastService.ts:109]: lets evaluate structures that allow you to use id as index, including keeping order such , https://www.npmjs.com/package/immutable

File: chat-components/src/services/BroadcastService.ts [hunk 6]
```diff
@@ -45,7 +51,18 @@ export const BroadcastService = {
          * Omit copying methods to prevent 'DataCloneError' in older browsers when passing an object with functions
          * This exception occurs when an object can't be clone with the 'structured clone algorithm' (used by postMessage)
          */
-        pubChannel.postMessage(JSON.parse(JSON.stringify(message)));
+        try {
+            const messageCopy = JSON.parse(JSON.stringify(message));
```
Comment (elopezanaya) [inline chat-components/src/services/BroadcastService.ts:55]: this line is prone to errors , should extract to a transform function? for safety

File: chat-components/src/services/BroadcastService.ts [hunk 7]
```diff
@@ -45,7 +51,18 @@ export const BroadcastService = {
...
+            const eventId = uuidv4();
+            const event = {...messageCopy, eventId};
+
+            eventQueue.pushEvent(event);
+            pubChannel.postMessage(event);
+        } catch (error) {
+            console.error("Error in BroadcastService.postMessage:", error);
+        }
+
+        eventQueue.processEvents(1500); // Second attempt to process events from queue if first try failed
```
Comment (elopezanaya) [inline chat-components/src/services/BroadcastService.ts:65]: 500 ms?, anything beyond 1 sec IMO is too much

File: chat-components/src/services/EventQueue.ts [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+    constructor(newMessage: Subject<ICustomEvent>) {
+        this.channelEventQueue = new Map<string, ICustomEvent>();
+        this.newMessage = newMessage;
+    }
+
+    processEvents(deferTimeout = 0) {
+        const dequeue = () => {
+            let queueSize = this.channelEventQueue.size; // Set queue size before processing to prevent infinite loop
+            while (queueSize > 0) {
```
Comment (elopezanaya) [inline chat-components/src/services/EventQueue.ts:18]: would be a foreach safer?

**Review by Erli-ms — APPROVED** [review]

**Review by elopezanaya — CHANGES_REQUESTED** [review]
comments from copilot , plus need to revisit timeout time for queueing events

File: chat-components/src/services/BroadcastService.ts [hunk 9]
```diff
@@ -35,17 +41,32 @@ export const BroadcastServiceInitialize = (channelName: string) => {
...
+            eventQueue.pushEvent(event);
+            pubChannel.postMessage(event);
+        } catch (error) {
+            console.error("Error in BroadcastService.postMessage:", error);
+        }
+
+        if (options?.retry) {
+            const deferTimeout = options?.deferTimeout || 500;
+            eventQueue.processEvents(deferTimeout); // Second attempt to process events from queue if first try failed
+        }
```
Comment (mithunjohny) [inline chat-components/src/services/BroadcastService.ts:69]: my thought is call eventQueue.queueEvents here which will start a timeout to invoke processEvents.

Comment (xTEddie) [reply chat-components/src/services/BroadcastService.ts]: Updated logic
